BUILD-92: added alerts, changes in degraded processing, for maintainability

gabemontero opened the pull request:
So this PR does a few things, separated by commits
1) adds alerts around failed imports and retries on imports using recently added metrics
2) cleans up the description for the missing secret alert (it got missed with the 4.5 change to no longer copy the pull secret)
3) and the biggie ... no longer report degraded if image stream imports are failing 

The feeling after living with this behavior for a bit is that it is too extreme ... yes, an "operand" is having some difficulty, but those problems arise from external image registries being problematic, not in cluster issues.

So we alert instead in order to notify administrators.

gabemontero commented:
```
error: some steps failed:
  * could not resolve inputs: could not determine inputs for step [input:ocp_4.0_base]: could not resolve base image: imagestreamtags.image.openshift.io "4.0" not found
time="2020-07-01T18:53:21Z" level=info msg="Reporting job state \"failed\" with reason \"resolving_inputs\""
```

maybe it is time to bump this up

gabemontero commented:
/retest

gabemontero commented:
/retest

gabemontero commented:
e2e's are hitting install not completing failures ... appears to be bzs opened but am tracking

I think we can get started with review

/assign @adambkaplan 

also,

/assign @bparees 

for historical perspective ... I've convinced @adambkaplan @sbose78 @otaviof @coreydaley that we should move off of marking samples operator degraded if the imagestream imports are failing

essentially, after a few releases of data, especially wrt registry.redhat.io stability, we are "redrawing the line" and will only go degraded if we

"failed to create our operands"

if after successful creation our operands have issues reaching 3rd party / external registries, we will only alert now

we will also alert if we do not have evidence that samples operator is periodically retrying failed imports.

thoughts?

bparees commented:
> essentially, after a few releases of data, especially wrt registry.redhat.io stability, we are "redrawing the line" and will only go degraded if we
"failed to create our operands"
if after successful creation our operands have issues reaching 3rd party / external registries, we will only alert now
we will also alert if we do not have evidence that samples operator is periodically retrying failed imports.
thoughts?

I'm torn.  We are degraded (we are not providing the service we are supposed to provide, namely, available imagestreams).  Heck, arguably we're not even "available". (our operand, the imagestreamtags, are not available). 

If the internal registry operator can't create S3 storage because AWS is down, it still marks itself unavailable because there's no usable internal registry deployed.

So in that sense, saying "we're not going to consider ourselves degraded/unavailable because it's due to circumstances outside our control" doesn't seem right.

On the other hand, this has been a huge pain point and the reality is that the samples operator being unable to import some content should not prevent upgrades from succeeding and in general is not a severe issue, so showing up degraded/unavailable due to it is causing undo concern for admins.

So from that perspective, i'm on board w/ limiting the cases where we go explicitly degraded/unavailable, and using alerts for the rest.

gabemontero commented:
> > essentially, after a few releases of data, especially wrt registry.redhat.io stability, we are "redrawing the line" and will only go degraded if we
> > "failed to create our operands"
> > if after successful creation our operands have issues reaching 3rd party / external registries, we will only alert now
> > we will also alert if we do not have evidence that samples operator is periodically retrying failed imports.
> > thoughts?
> 
> I'm torn. We are degraded (we are not providing the service we are supposed to provide, namely, available imagestreams). Heck, arguably we're not even "available". (our operand, the imagestreamtags, are not available).
> 
> If the internal registry operator can't create S3 storage because AWS is down, it still marks itself unavailable because there's no usable internal registry deployed.
> 
> So in that sense, saying "we're not going to consider ourselves degraded/unavailable because it's due to circumstances outside our control" doesn't seem right.
> 
> On the other hand, this has been a huge pain point and the reality is that the samples operator being unable to import some content should not prevent upgrades from succeeding and in general is not a severe issue, so showing up degraded/unavailable due to it is causing undo concern for admins.
> 
> So from that perspective, i'm on board w/ limiting the cases where we go explicitly degraded/unavailable, and using alerts for the rest.

You've concisely encapsulated my evolution on this over the past 6 to 12 months.

Thanks for the quick input.  We'll move forward on this.

gabemontero commented:
/retest

sbose78 commented:
Sounds good @gabemontero .

> we should move off of marking samples operator degraded if the imagestream imports are failing

+1 

> samples operator being unable to import some content should not prevent upgrades from succeeding and in general is not a severe issue, so showing up degraded/unavailable due to it is causing undo concern for admins.

+1 

Yeah, in the scenario where samples operator is unable to import, we are degraded **but** not really degraded enough to be a show-stopper for bothering admins.

gabemontero commented:
/retest

adambkaplan commented:
> On the other hand, this has been a huge pain point and the reality is that the samples operator being unable to import some content should not prevent upgrades from succeeding and in general is not a severe issue, so showing up degraded/unavailable due to it is causing undo concern for admins.

IMO blocking upgrade is the biggest concern with our current approach. Plus, with the alert the cluster admin is still aware that not all samples are available. From our telemetry and CI, in many cases samples are partially available.

bparees commented:
> IMO blocking upgrade is the biggest concern with our current approach. 

yeah that's what ultimately pushed me over the edge on supporting this change

adambkaplan commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,30 @@ spec:
+        severity: warning
+      annotations:
+        message: |
+          Samples operator is detecting problems with imagestream image imports, and the periodic retries of those imports are not occurring.  Contact support.  You can look at the "openshift-samples" ClusterOperator object for details. You can also query 'openshift_samples_failed_imagestream_import_info' to see which ImageStreams do not import. Most likely there are issues with the external imag ...
```
Could we use alerting rule templates here [1]? Could we surface the individual ImageStreams that are failing to import?

I am a bit loathe to instruct admins to query a metric when the alerting rules have access to the same info.

[1] https://prometheus.io/docs/prometheus/latest/configuration/template_examples/

gabemontero replied:
The surface individual imagestreams with the underlying metric and hence corresponding alert was an earlier discussion point of ours as I'm sure you remember, where we went with coarse grained / do not include specific imagestream(s).

Not sure I'm inclined to change course on that yet.  Among other things, I'm not sure exactly how to do it yet.

I'll read up on templates in the interim.  Feel free to elaborate on the "how" while I do that.

gabemontero replied:
So I spent a day trying to produce an alert @adambkaplan with the current form of the metric / query but with no luck

I think I have a new query which should work.  Will try that out tomorrow.  Assuming I have luck, I'll then see about changing the new retry metric to a CounterVec to see about adding labels, etc. for use in templates.

adambkaplan commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,30 @@ spec:
       annotations:
         message: |
-          Samples could not be deployed and are in Degraded status. You can look at the "openshift-samples" ClusterOperator object for details. You can also query 'openshift_samples_failed_imagestream_import_info' to see if ImageStreams do not import. The samples operator reports itself Degraded if an ImageStream import continues to fail for 2 hours.
+          Samples could not be deployed and are in Degraded status. You can look at the "openshift-samples" ClusterOperator object for details.
```
Avoid the second-person - use an imperative instead:
```suggestion
          Samples could not be deployed and the operator degraded. Review the "openshift-samples" ClusterOperator object for further details.
```

gabemontero replied:
part of next push

adambkaplan commented on pkg/metrics/metrics.go:
```diff
@@ -49,7 +49,7 @@ var (
 		nil)
 	invalidSecretDesc = prometheus.NewDesc(invalidSecretQuery,
-		"Indicates if the pull secret is functional in the openshift namespace.  If not functional, the reason supplied either points to it missing in the openshift namesapce, or if it is missing credentials for registry.redhat.io when they should exist (1 == not functional, 0 == functional).",
+		"Indicates if the install pull secret is functional for pulling images from registry.redhat.io.  If not functional, the reason supplied either points to it being inaccessible, or if it is missing credentials for registry.redhat.io when they should exist (1 == not functional, 0 == functional).",
```
Can we replace "not functional" with "invalid"?

gabemontero replied:
part of next push

gabemontero commented:
/retest

gabemontero commented:
@adambkaplan I've pushed changes in a separate commit for your two text based comments

gabemontero commented:
for historical reference  .... install errors like 

```
level=error msg="Cluster operator monitoring Degraded is True with UpdatingPrometheusK8SFailed: Failed to rollout the stack. Error: running task Updating Prometheus-k8s failed: waiting for Prometheus object changes failed: waiting for Prometheus openshift-monitoring/k8s: expected 2 replicas, got 0 updated replicas"
```

can stem from invalid PrometheusRule definitions in your PR's set of changes :-)

gabemontero commented:
/test e2e-aws

gabemontero commented:
following up from https://github.com/openshift/cluster-samples-operator/pull/300#discussion_r451175760 I've got the alerts working like I would expect with a single dimension (i.e. no imagestream name labels) using the new retry metric.

tl;dr - and/or in alert expr's are not as intuitive as one would think (basically only work in the context of vectors)

I'll take a stab, in a separate commit, at making the new metric a counter vector, adding imagestream name labels, and using templates to spit out imagestream names in the alert per the ask from @adambkaplan starting tomorrow. We can compare the deltas and decide which way we want to settle on.

gabemontero commented:
/test unit

gabemontero commented:
/test e2e-aws-operator

gabemontero commented:
/retest

gabemontero commented:
ok @adambkaplan I've added a commit that 
- changes the new retry metric to a CounterVec
- adds an imagestreamname label
- updated the alert query to accommodate (and successfully tested locally)
- updated the alert message to use prom/golang templates and include imagestreams of note in the new alerts

PTAL

Based on which way we go, I'll arrange / squash commits accordingly.

gabemontero commented:
https://projects.engineering.redhat.com/browse/RHELPLAN-48796 has been opened for the new django template issue in image eco.
Revert PR https://github.com/sclorg/django-ex/pull/154 is a possible workaround, though we'll give the upstream devs some time to sort it out.

gabemontero commented:
/test e2e-aws-upgrade

gabemontero commented:
/test e2e-aws-upgrade

gabemontero commented:
/retest

paulfantom commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,48 @@ spec:
+          imports are not occurring.  Contact support.  You can look at the "openshift-samples" ClusterOperator object
+          for details. Most likely there are issues with the external image registry hosting the images that need to
+          be investigated.  The list of ImageStreams that have failing imports are:
+          {{ range query "openshift_samples_retry_imagestream_import_total > 0" }}
+            {{ .Labels.name }} {{ .Value }}
+          {{ end }}
+          However, the list of ImageStreams for which samples operator is retrying imports is:
+          retrying imports:
+          {{ range query "openshift_samples_retry_imagestream_import_total > 0" }}
+             {{ .Labels.imagestreamname }} {{ .Value }}
+          {{ end }}
```
I don't think this will work as this might not be read as a template. Have you tested it?

gabemontero replied:
The alert fired fine when I looked at it from the prometheus UI, but I have yet tried to look at it from the alert manager UI.  I'm going to try and do that today and report back.  I'm assuming if I don't see the expected var substitution there that would confirm your suspicion.

Thanks for the quick feedback @paulfantom

gabemontero replied:
Also, one quick ask for clarification.  Looking at https://prometheus.io/docs/prometheus/latest/configuration/template_examples/#simple-alert-field-templates is the difference in that example and this that the example in the doc deals with the description, where as I'm trying to use templates for the message @paulfantom ?

gabemontero replied:
It looks like it works @paulfantom .... from the alertmanager view from the console:

```
Samples operator is detecting problems with imagestream image imports. You can look at the "openshift-samples" ClusterOperator object for details. Most likely there are issues with the external image registry hosting the images that needs to be investigated. Or you can consider marking samples opertaor Removed if you do not care about having sample imagestreams available. The list of ImageStreams for which samples operator is retrying imports: apicurito-ui 4 dotnet 4 eap-cd-openshift 17 fuse7-console 6 fuse7-karaf-openshift 12 httpd 1 java 2 jboss-amq-62 14 jboss-amq-63 6 jboss-datagrid65-client-openshift 2 jboss-datagrid65-openshift 5 jboss-datagrid71-client-openshift 3 jboss-datagrid72-openshift 6 jboss-datagrid73-openshift 5 jboss-datavirt64-openshift 8 jboss-decisionserver64-openshift 7 jboss-eap71-openshift 5 jboss-eap72-openshift 4 jboss-fuse70-eap-openshift 2 jboss-processserver64-openshift 7 jboss-webserver30-tomcat8-openshift 5 nginx 3 nodejs 3 openjdk-11-rhel7 4 openjdk-11-rhel8 3 openjdk-8-rhel8 4 perl 1 php 2 redhat-sso70-openshift 2 redhat-sso71-openshift 5 redhat-sso72-openshift 10 rhpam-businesscentral-rhel8 3 apicast-gateway 18 dotnet-runtime 2 eap-cd-runtime-openshift 2 fis-java-openshift 2 fis-karaf-openshift 4 fuse-apicurito-generator 4 fuse7-eap-openshift 6 fuse7-java-openshift 6 golang 2 jboss-datagrid71-openshift 4 jboss-datavirt64-driver-openshift 8 jboss-eap64-openshift 10 jboss-eap70-openshift 5 jboss-fuse70-console 1 jboss-fuse70-java-openshift 1 jboss-fuse70-karaf-openshift 3 jboss-webserver30-tomcat7-openshift 6 jboss-webserver31-tomcat7-openshift 5 jboss-webserver31-tomcat8-openshift 5 jboss-webserver50-tomcat9-openshift 8 mariadb 2 mongodb 5 mysql 4 postgresql 3 python 4 redhat-openjdk18-openshift 9 redhat-sso73-openshift 2 redis 2 rhdm-decisioncentral-rhel8 1 rhdm-kieserver-rhel8 1 rhpam-businesscentral-monitoring-rhel8 2 rhpam-kieserver-rhel8 2 rhpam-smartrouter-rhel8 1 ruby 3 sso74-openshift-rhel8
```

@adambkaplan FYI ^^

gabemontero replied:
Though I'm not sure where those seemingly random numbers come from ... each imagestream name is followed by a number i.e. 

`httpd 1 java 2`

investigating

gabemontero replied:
figured it out .... will push update momentarily ... new output 

Samples operator is detecting problems with imagestream image imports. You can look at the "openshift-samples" ClusterOperator object for details. Most likely there are issues with the external image registry hosting the images that needs to be investigated. Or you can consider marking samples opertaor Removed if you do not care about having sample imagestreams available. The list of ImageStreams for which samples operator is retrying imports: apicurito-ui dotnet eap-cd-openshift fuse7-console fuse7-karaf-openshift httpd java jboss-amq-62 jboss-amq-63 jboss-datagrid65-client-openshift jboss-datagrid65-openshift jboss-datagrid71-client-openshift jboss-datagrid72-openshift jboss-datagrid73-openshift jboss-datavirt64-openshift jboss-decisionserver64-openshift jboss-eap71-openshift jboss-eap72-openshift jboss-fuse70-eap-openshift jboss-processserver64-openshift jboss-webserver30-tomcat8-openshift nginx nodejs openjdk-11-rhel7 openjdk-11-rhel8 openjdk-8-rhel8 perl php redhat-sso70-openshift redhat-sso71-openshift redhat-sso72-openshift rhpam-businesscentral-rhel8 apicast-gateway dotnet-runtime eap-cd-runtime-openshift fis-java-openshift fis-karaf-openshift fuse-apicurito-generator fuse7-eap-openshift fuse7-java-openshift golang jboss-datagrid71-openshift jboss-datavirt64-driver-openshift jboss-eap64-openshift jboss-eap70-openshift jboss-fuse70-console jboss-fuse70-java-openshift jboss-fuse70-karaf-openshift jboss-webserver30-tomcat7-openshift jboss-webserver31-tomcat7-openshift jboss-webserver31-tomcat8-openshift jboss-webserver50-tomcat9-openshift mariadb mongodb mysql postgresql python redhat-openjdk18-openshift redhat-sso73-openshift redis rhdm-decisioncentral-rhel8 rhdm-kieserver-rhel8 rhpam-businesscentral-monitoring-rhel8 rhpam-kieserver-rhel8 rhpam-smartrouter-rhel8 ruby sso74-openshift-rhel8

paulfantom commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,48 @@ spec:
+          {{ range query "openshift_samples_retry_imagestream_import_total > 0" }}
+             {{ .Labels.imagestreamname }} {{ .Value }}
+          {{ end }}
+    - alert: SamplesImagestreamImportFailing
```
This looks very similar to `SamplesRetriesMissingOnImagestreamImportFailing`. Won't they fire at the same time in most cases? What are the cases when only one of them fires?

gabemontero replied:
No they do not fire at the same time.  Testing has confirmed.
- `SamplesRetriesMissingOnImagestreamImportFailing` fires when imagestream imports are failing but for whatever reason the samples operator is *NOT* attempting to retry imports
- `SamplesImagestreamImportFailing` fires when imagestream imports are failing *AND* samples operator is retrying as expected

gabemontero commented:
/retest

adambkaplan commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,48 @@ spec:
+          {{ range query "openshift_samples_retry_imagestream_import_total > 0" }}
+            {{ .Labels.name }} {{ .Value }}
+          {{ end }}
+          However, the list of ImageStreams for which samples operator is retrying imports is:
+          retrying imports:
+          {{ range query "openshift_samples_retry_imagestream_import_total > 0" }}
+             {{ .Labels.imagestreamname }}
+          {{ end }}
```
Looks like this will report the same set of imagestreams, no? I don't think we need this list.

gabemontero replied:
yeah there was suppose to be lists from two separate metrics ... fixed in latest push - thanks

adambkaplan commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,48 @@ spec:
       annotations:
         message: |
-          Samples could not be deployed and are in Degraded status. You can look at the "openshift-samples" ClusterOperator object for details. You can also query 'openshift_samples_failed_imagestream_import_info' to see if ImageStreams do not import. The samples operator reports itself Degraded if an ImageStream import continues to fail for 2 hours.
+          Samples could not be deployed and the operator degraded. Review the "openshift-samples" ClusterOperator object for further details.
```
Grammar nit:
```suggestion
          Samples could not be deployed and the operator is degraded. Review the "openshift-samples" ClusterOperator object for further details.
```

adambkaplan commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -40,7 +74,10 @@ spec:
       annotations:
         message: |
-          Samples operator cannot find credentials for 'registry.redhat.io' in the samples pull secret in the openshift namespace.  Unless you have updated the 'samplesRegistry' field in the samples operator config to change which registry the sample ImageStreams are pulled from, you will have failed imports for many of the sample ImageStreams in this case.
+          Samples operator cannot find credentials for 'registry.redhat.io' in the samples pull secret in the openshift
+          namespace.  Unless you have updated the 'samplesRegistry' field in the samples operator config to change which
+          registry the sample ImageStreams are pulled from, you will have failed imports for many of the sample ImageStreams
+          in this case.
```
I likewise think we make this more concise (even though this was an existing alert):
```suggestion
          The samples operator cannot find credentials for 'registry.redhat.io'. Many of the sample ImageStreams will fail to import unless the 'samplesRegistry' in the operator configuration is changed.
```

adambkaplan commented on pkg/metrics/server.go:
```diff
@@ -82,6 +83,6 @@ func TBRInaccessibleOnBoot(badTBR bool) {
 
-func ImageStreamImportRetry() {
-	importRetryStat.Inc()
+func ImageStreamImportRetry(imagestreanName string) {
+	importRetryStat.With(prometheus.Labels{"imagestreamname": imagestreanName}).Inc()
```
nit:
```suggestion
func ImageStreamImportRetry(imagestreamName string) {
	importRetryStat.With(prometheus.Labels{"imagestreamname": imagestreamName}).Inc()
```

adambkaplan commented on manifests/010-prometheus-rules.yaml:
```diff
@@ -9,14 +9,48 @@ spec:
+          {{ range query "openshift_samples_failed_imagestream_import_info > 0" }}
+            {{ .Labels.name }}
+          {{ end }}
+          However, the list of ImageStreams for which samples operator is retrying imports is:
+          retrying imports:
```
nit:
```suggestion
          However, the list of ImageStreams for which samples operator is retrying imports is:
```

gabemontero replied:
Hmm ... the commit suggestion button is greyed out @adambkaplan and when I hover my mouse over the commit suggestion button it says "This suggestion is invalid because no changes were made to the code."

adambkaplan replied:
suggestion may have come from a prior commit. I think this is good.

gabemontero commented:
all green tests as well @adambkaplan 

aside from figuring out what you were trying to convey with https://github.com/openshift/cluster-samples-operator/pull/300#discussion_r453858701 (or was it already addressed maybe?) perhaps we are in squash and merge mode, unless you take another pass and provide more input as a result

gabemontero commented:
thanks @adambkaplan 

commits squash / commit message updated and pushed

gabemontero commented:
samples operator upgrade fine in https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-samples-operator/300/pull-ci-openshift-cluster-samples-operator-master-e2e-aws-upgrade/1283091324970471424 but other operators failed

gabemontero commented:
/retest

gabemontero commented:
all green tests and commits squashed @adambkaplan

adambkaplan commented:
/retitle Alert if imagestream imports are retrying, instead of reporting Degraded
